tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -4337,13 +4378,22 @@ class CudnnExecutionPlanRunner<void(Args...)>
 
     std::array<void*, sizeof...(Args)> data_ptrs = {inputs.opaque()...};
 
-    // TODO(kaixih@nvidia): Remove the const_cast after cudnn frontend can take
-    // in const int64 pointer.
+    absl::InlinedVector<int64_t, sizeof...(Args)> data_uids_vec(
+        data_uids_.cbegin(), data_uids_.cend());
+    absl::InlinedVector<void*, sizeof...(Args)> data_ptrs_vec(
+        data_ptrs.cbegin(), data_ptrs.cend());
+    // We use the need_side_input to determine if the side input 'z' from
+    // {'x', 'w', 'z', 'b', 'y'} is needed for the conv-<add>-bias-act patterns.
+    if (sizeof...(Args) == 5 && !need_side_input_) {
+      data_uids_vec.erase(data_uids_vec.begin() + 2);
+      data_ptrs_vec.erase(data_ptrs_vec.begin() + 2);
+    }
```

Comment:
If I am following this code correctly...

We pass {'x', 'w', 'z', 'b', 'y'} and need_side_input=false to this function, and then if need_side_input == false, we remove 'z' from the array.

That's a hack, right?  We should instead be passing {'x', 'w', 'b', 'y'} from the beginning?

Reply:
Yes, this is a hack. This is a known issue for cudnn. Cudnn should have supported this pattern Conv-Add-BiasAdd-Act no matter what the Act is. But for now, cudnn requires the Act to be Relu, otherwise we have to use Conv-BiasAdd-Act pattern. I tried to work around this by using less code. If we instead pass {'x', 'w', 'b', 'y'}, we need to create new FusedConvSignature in dnn.h and create new objects for the CudnnExecutionPlanRunner, and plumbing all these together.

Reply:
Oh, I see, the number of params here has to match all the way up to FusedConvSignature in dnn.h.  Yikes.

I think adding a third conv signature is kind of the correct way to do it, though?  Today we have ConvSignature and FusedConvSignature, which take 3 and 5 arguments, respectively.  We're adding a new fusion kind that actually only uses four arguments, but we're kind of shoving it into FusedConvSignature, saying that we'll ignore one of the args.

That doesn't seem to match the spirit of the design?  Like, if we are going to do that, we might as well make *everything* use the 5-arg version.  Which might also be okay, I am not sure how bad that would be.  But having 3- and 5-arg signatures and no 4-arg signature seems not good?

Reply:
Emm, if we really want a clean conv-bias-act, we can propose the 4-arg signature. But I think for this PR the 5-arg case can already cover what all this PR wants to do (though it's a bit ugly.). 

I understand your concern for ignoring one of the args. But I think: (1) this is a WAR for cudnn's failure to pick the correct path for Conv-Add-Bias-NonRelu (and we have to use Conv-Bias-NonRelu to trigger the right path); (2), I notice that the native TF always use scales 1 and 0 for the Add node, meaning the Add is essentially a Noop. Maybe XLA uses different scales. So if that is the case, I can add more checks on the scales to make sure a safe drop of Add node.

Reply:
> (1) this is a WAR for cudnn's failure to pick the correct path for Conv-Add-Bias-NonRelu (and we have to use Conv-Bias-NonRelu to trigger the right path)

Oh, I think I finally understand.  You are saying that, basically, we *want* to use conv-add-bias-nonrelu and pass 0 for the addend scale (thus making this identical to the relu case)?  But we can't do that due to the cudnn bug?

In that case, sure, if we add a CHECK that the add-scale is always 0 for nonrelu activation then this feels safe and not as bad of a hack to me.

Reply:
> Oh, I think I finally understand. You are saying that, basically, we want to use conv-add-bias-nonrelu and pass 0 for the addend scale (thus making this identical to the relu case)? But we can't do that due to the cudnn bug?
Yes, exactly.